### ISSUE TITLE
Please provide a clean method to suppress layout.

### ISSUE THREAD

**Issue by gifad** [backdrop/backdrop-issues#1728]
I need a method to show pages with "content only".

A hack based on `$_SERVER['HTTP_ACCEPT']` has been suggested in https://github.com/backdrop/backdrop-issues/issues/1483#issuecomment-169173412 but it won't work in 1.3.4 with https://github.com/backdrop/backdrop/pull/1290

This is currently used in my port of [References dialog](https://github.com/backdrop-contrib/references_dialog)
Do I have to commit another hack based on a fake `$_SERVER['HTTP_X_REQUESTED_WITH']` ?

A `function layout_suppress()`, similar to `admin_bar_suppress()` would be convenient.
(just to request `layout_route_handler()` tu use `menu_default_route_handler()`)

Note: the drupal version uses a query `?render=references-dialog` to fire a page template...

**Comment by gifad** [comment]
Can this be qualified as "clean" ?
```
function references_dialog_init() {
  if (references_dialog_in_dialog()) {
    admin_bar_suppress(TRUE);
    $config = config('system.core');
    $config->set('menu_route_handler', 'menu_default_route_handler');
```

**Comment by gifad** [comment]
Otherwise, please consider PR [#1291 Introduce layout_suppress()](https://github.com/backdrop/backdrop/pull/1291)

**Comment by quicksketch** [comment]
This might be related to https://github.com/backdrop/backdrop-issues/issues/1664, where we're talking about switching from basing responses from Accept headers to using query strings. If we took that approach, it would probably alleviate the need for this work-around. Though this could probably done as well.

@docwilmot do you have any thoughts on this suggestion? You've been looking into Layouts as much as anyone.

**Comment by docwilmot** [comment]
I dont know much about #1664 TBH, but this PR here sounds like a good idea to me. I can see how it may be more convenient than trying to force a layout.

Just looking at the PR though, shouldnt the implementation be 
```
  // If requested by layout_suppress(), use the default route handler instead of layouts.
  if (layout_suppress()) {
    return menu_default_route_handler($router_item);
  }
```

seems `layout_suppress(FALSE)` would *not* suppress, no?

**Comment by gifad** [comment]
The signature and behavior of `layout_suppress()` has been copied verbatim from `admin_bar_suppress()`, which is very similar. I won't comment these (I don't know of "correct" english words to do it), so I preferred to stay consistent.

Feel free to provide a better suggestion, I just need to choose a fix for References dialog with 1.3.4.

**Comment by quicksketch** [comment]
Yeah, I can see what you were going for here, matching admin_bar. The code in `admin_bar_suppress()` seems to be only callable once (you can't un-supress). I don't think we have that same limitation here. Instead of a `$set` parameter, let's make it `$suppress` and take an optional boolean TRUE/FALSE to set the suppression state. The current suppression state will always be returned. This would make it behave in the same manner as `backdrop_page_is_cacheable()`.

With @docwilmot's approval, I'm okay with this addition as well. Let's just make the above changes and then we can put it in.

**Comment by klonos** [comment]
Since @gifad seems to have removed their GitHub account, I have created a new PR here, based off of his: https://github.com/backdrop/backdrop/pull/2891

@gifad had already addressed @quicksketch's concerns, but did not update this ticket to let us know. So I have resolved conflicts with 1.x, and have additionally merged the conditional with the one directly above it (they were both returning the same thing).

Should we merge this and close this oldie? ...it's just a small addition 😉

**Comment by olafgrabienski** [comment]
@klonos Thanks for making the PR: I've just found it when I started to use the *References Dialog* module and stumbled about issue https://github.com/backdrop-contrib/references_dialog/issues/4.

I've tested the changes of this PR locally, in combination with uncommenting [one line in References Dialog](https://github.com/backdrop-contrib/references_dialog/blob/1.x-1.x/references_dialog.module#L457)). And indeed, that fixes the issue for me!

However, there went some changes into `layout.module` since september 2019. Can you update the PR once more?

**Comment by indigoxela** [comment]
I think, we should provide a (relatively) easy way to suppress the layout and only provide the content. That's something really complex with layouts.

@olafgrabienski One problem I see: this issue is marked as "feature request", which means it can not get into core before January next year.

**Comment by olafgrabienski** [comment]
> this issue is marked as "feature request", which means it can not get into core before January next year.

No problem, I can wait (and use a patch in the meantime). Or could it also be considered as task?

**Comment by indigoxela** [comment]
I've left a question on the PR. Unfortunately @gifad isn't around anymore to explain, so maybe someone else can. If so, please do.

**Comment by jenlampton** [comment]
Other than the issue with references_dialog (which looks like it may have a simple fix?) do we have any valid use cases for this change? Before we add a new feature to core, I'd like to be sure it meets the 80% use-case (of it it is needed to make contrib alternatives possible -- has there been any attempt to do this in contrib?)

I worry when I see an issue like this that it's something that may be requested by people new to Backdrop who aren't fully aware of how Layouts work yet, and once they are Backdrop experts they'll know other ways to achieve what they need. 

And if so, is it still valuable to support something that "makes it work" for the novice, even though there are already other/better/safer ways to achieve what they need?

**Comment by indigoxela** [comment]
> Other than the issue with references_dialog (which looks like it may have a simple fix?)

It doesn't have a simple fix. To my understanding using the ajax dialog requires a lot of rewrite.

> I'd like to be sure it meets the 80% use-case

It's not a 80% use-case but it's something that was easy with Drupal 7 and isn't easy at all with Backdrop.

> is it still valuable to support something that "makes it work" for the novice, even though there are already other/better/safer ways to achieve what they need?

There aren't easy ways to achieve that. That's why I support adding this simple method.
Honestly, what's wrong with providing a simple way to skip the layout wrapper? I don't get the problem. :confused:
Why sticking with a contrib-breaker when it's so easy to provide a solution?

**Comment by jenlampton** [comment]
> It doesn't have a simple fix. To my understanding using the ajax dialog requires a lot of rewrite.

Nate spotted a tiny mistake in the module that was causing the problem (he mentioned it in the weekly meeting today -- I'm not sure if he's gotten around to posting a comment yet, but he did a screenshare where he pointed out what needed to be fixed).

**Comment by indigoxela** [comment]
Does this mean, this PR is rejected? Or is the whole issue rejected? Please explain.

> I'm not familiar enough with layouts to answer why this approach is not recommended

If it's the approach - would a different approach have any chance? And why is there such a resistance against an easy way for layout-less output? It doesn't have to be like the currently available PR. Can @docwilmot explain the "danger"?

**Comment by jenlampton** [comment]
> And why is there such a resistance against an easy way for layout-less output? 

My resistance is only because we already have an easy way for layout-less output in core.

**Comment by indigoxela** [comment]
> My resistance is only because we already have an easy way for layout-less output in core.

OK, can you give some simple examples, please?

**Comment by jenlampton** [comment]
Try this? https://youtu.be/vM0tIjF6Wgk?t=390 (my youtube link skills are not great...)

**Comment by indigoxela** [comment]
> Try this?

A link to the weekly meeting? Can't you simply provide some lines of code as example?

**Comment by indigoxela** [comment]
OK, the "simple" solution is to add a specially crafted $_SERVER variable in an ajax POST request? Not impressed.

Anyway, giving up here. :wink:

**Comment by quicksketch** [comment]
Sorry I started posting a comment here earlier but didn't finish it before weekly meetings. What References Dialog had been doing in the past (and currently) is manipulating the global `$_SERVER` variables to suppress layouts. What it should have been doing instead is making an HTTP request that specifies headers that Backdrop considers to be a non-typical HTTP request.

Line [63 of `references-dialog.js`](https://github.com/backdrop-contrib/references_dialog/blob/1.x-1.x/js/references-dialog.js#L63) is doing the following:

```
    $.post(callback, entity_info);
```

But it's not specifying any HTTP headers at all. If it were to change to this:

```
$.ajax({
  url: callback,
  context: entity_info,
  headers: {
    accept: 'text/x-references-dialog'
  },
);
```

Then I think that would make `backdrop_is_html()` return `FALSE` and the normal page callback result would be returned, without using a layout.

Even with that current solution available (if it works), I still think adding this might be a good idea. Who knows what other situations you may need to disable layouts? I think what I was trying to say in the meeting was that this particular use-case there may already be a better solution for, but in general this could be valuable anyway.

**Comment by indigoxela** [comment]
@quicksketch many thanks for that example (also linked it in the module issue, so they can proceed).

Now back to core. :wink: 

> ... that would make backdrop_is_html() return FALSE...

This feels sort of hackish - it *is* html, claiming that it isn't, only to get rid of the layout wrappers seems odd to me. (Or at least, the function name is misleading then.)

Another thought: it's relatively easy to bypass the layout wrappers when using links with special attributes (`class="use-ajax" data-dialog="true"`), but currently there's no obvious way to do it without. (And even this link approach isn't known by everyone.)

And I don't see benefit in making it so hard to skip the layout wrappers.

If it's preferred to stick with custom http accept headers, it seems essential to me to document that in a way, that it's easy to find and easy to implement.

**Comment by herbdool** [comment]
I've fixed the PR a bit. In my testing it works.

I think most people find this useful, but just wanted to add another possible use case. It might help with https://github.com/backdrop/backdrop-issues/issues/4410 - perhaps have an option to just suppress the layout altogether on the login page.

BTW, setting the header in `$.ajax` didn't change anything for me, so might require a bit more investigation for that approach.

**Comment by olafgrabienski** [comment]
@herbdool Thanks for updating the PR. I've applied the changes locally, they work for me as well.

I guess however, the PR has to be rebased (is this the right term?), because `layout.module` changed since PR creation in 2019 by @klonos. E.g, you find `return call_user_func_array('_layout_page_access', $args);` which is directly before the main PR change in [line 536 of the PR](https://github.com/backdrop/backdrop/pull/2891/files#diff-82f95b6e5ad80e32e562b32b07e58cff1770943d5761e270eeef770c0fb55339R536), but in the meantime it moved to [L577](https://github.com/backdrop/backdrop/blob/1.x/core/modules/layout/layout.module#L577).

**Comment by klonos** [comment]
> I guess however, the PR has to be rebased...

Yes. Sorry that you had to ask twice @olafgrabienski. I just rebased the PR, and now Tugboat is also happy with it.

**Comment by indigoxela** [comment]
@klonos many thanks for rebasing.

I also tested locally with following minimal code in a custom module:
```
function testthing_init() {
  $params = backdrop_get_query_parameters();
  if (isset($params['layout-go-away'])) {
    layout_suppress(TRUE);
  }
}
```
An url like `/node/add/post?layout-go-away` isn't wrapped in a layout, while `/node/add/post` still is - works for me.

The code is straight forward, I'm happy to see this going into 1.21.0.
Unfortunately this means that people who need modules like References dialog will have to patch core until January. (The alternate approach suggested in a previous comment turned out to not work.)

RTBC!

**Comment by jenlampton** [comment]
Rather than patching core to add a new feature, a new release of references
dialog that fixes its bug would be a better solution ;)

**Comment by herbdool** [comment]
Most people in the thread see this as a useful addition to core.

**Comment by jenlampton** [comment]
Yes, but that's no reason to leave a contrib module broken for 4 months ;)
